# Feature Engineering

The four findings from `01_eda.ipynb` set the agenda for this notebook:

1. Monetary variables are heavy-tailed and need log transformation.
2. Fraud is structurally restricted to `TRANSFER` and `CASH_OUT`, so transaction type must be encoded explicitly.
3. The accounting residual `oldbalanceOrg − amount − newbalanceOrig` separates fraud from legitimate activity, partly as a simulator artefact.
4. No single variable carries the signal.

We construct the corresponding features below, perform the train/test split before any scaling so the test set never influences preprocessing, and persist the artefacts that the modelling notebooks and the FastAPI prototype need.


In [1]:
from pathlib import Path

import joblib
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import RobustScaler, StandardScaler

PROCESSED_DIR = Path("..") / "data" / "processed"
MODELS_DIR = Path("..") / "src" / "models"
FIGURES_DIR = Path("..") / "reports" / "figures"
RAW_PATH = Path("..") / "data" / "raw" / "PS_20174392719_1491204439457_log.csv"

PROCESSED_DIR.mkdir(parents=True, exist_ok=True)
MODELS_DIR.mkdir(parents=True, exist_ok=True)
FIGURES_DIR.mkdir(parents=True, exist_ok=True)

RANDOM_STATE = 42
TEST_SIZE = 0.2
sns.set_theme(style="whitegrid", context="notebook")

In [2]:
df = pd.read_csv(RAW_PATH)
df.shape

(6362620, 11)

## 1. Train/test split before feature engineering

Any statistic computed across the full dataset means, standard deviations, encoders fit on the union of train and test, leaks information from the test set into preprocessing. We split first and treat the test set as if it does not exist for the rest of this notebook. The scaler is fit on the training set only, then applied to both.


In [3]:
# Identifier columns are not predictive and would only inflate dimensionality
# if encoded. isFlaggedFraud is a hard-coded rule on the same target signal we
# are trying to predict, so it would act as label leakage if used as a feature.
columns_to_remove = ["nameOrig", "nameDest", "isFlaggedFraud"]
df = df.drop(columns=columns_to_remove)

X = df.drop(columns=["isFraud"])
y = df["isFraud"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=TEST_SIZE,
    stratify=y,
    random_state=RANDOM_STATE,
)

print(f"Train: {X_train.shape}, fraud rate {y_train.mean():.4%}")
print(f"Test:  {X_test.shape}, fraud rate {y_test.mean():.4%}")

Train: (5090096, 7), fraud rate 0.1291%
Test:  (1272524, 7), fraud rate 0.1291%


## 2. Feature construction

We define the full transformation as a single function. There are two reasons for this. First, the same logic must run at inference time inside the FastAPI service, where we score one transaction at a time, putting the logic in a function makes it trivial to lift into `src/features/build_features.py`. Second, applying it identically to train and test rules out the class of bugs where the two go through subtly different code paths.

The function takes a copy and returns a new dataframe. No in-place mutation, no surprising side-effects on the caller.


In [4]:
def build_features(frame: pd.DataFrame) -> pd.DataFrame:
    """
    Construct the engineered feature set used by the supervised baselines and
    the anomaly detectors.

    Features are organised into four groups:

    - **Magnitude.** ``log_amount`` absorbs the heavy-tailed structure of
      transaction amounts identified in the EDA.
    - **Balance residuals.** ``balance_error`` is the residual from the
      identity ``newbalanceOrig = oldbalanceOrg - amount`` and was the
      strongest behavioural signal in the EDA. ``abs_balance_error`` and
      ``log_balance_error`` give scale-stable variants.
    - **Interaction.** ``log_amount_balance_error`` couples transaction
      magnitude with the absolute balance residual. The log keeps it on a
      comparable scale to the other features and avoids the explosive
      magnitude of the raw product.
    - **Destination and ratio.** ``dest_balance_change`` captures the
      receiver's reaction to the transaction; ``log_amount_to_balance_ratio``
      captures how large the transaction is relative to the sender's
      available funds.

    Transaction type is one-hot encoded with ``dtype=int`` so the columns are
    not silently excluded when downstream code uses
    ``select_dtypes(include=[np.number])`` — in modern pandas
    ``pd.get_dummies`` returns ``bool`` by default, which is not in
    ``np.number``.

    Parameters
    ----------
    frame : pandas.DataFrame
        Raw PaySim transactions. Must contain ``amount``, ``oldbalanceOrg``,
        ``newbalanceOrig``, ``oldbalanceDest``, ``newbalanceDest`` and
        ``type``.

    Returns
    -------
    pandas.DataFrame
        Engineered feature matrix. The original raw columns are dropped; the
        caller is responsible for re-attaching the target if needed.
    """
    out = frame.copy()

    # --- magnitude --------------------------------------------------------
    out["log_amount"] = np.log1p(out["amount"])

    # --- balance residuals -----------------------------------------------
    # Note: the EDA showed this residual is essentially zero for fraud and
    # broadly dispersed for legitimate activity. Useful as a feature, partly
    # an artefact of the simulator — see the discussion in 01_eda.ipynb.
    balance_error = out["oldbalanceOrg"] - out["amount"] - out["newbalanceOrig"]
    out["abs_balance_error"] = balance_error.abs()
    out["log_balance_error"] = np.log1p(out["abs_balance_error"])

    # --- interaction term ------------------------------------------------
    out["log_amount_balance_error"] = np.log1p(
        out["amount"] * out["abs_balance_error"]
    )

    # --- ratio and destination ------------------------------------------
    # Adding a small constant prevents division by exact zero, which is
    # common in PaySim where many sender balances are 0 by construction.
    epsilon = 1e-6
    out["log_amount_to_balance_ratio"] = np.log1p(
        out["amount"] / (out["oldbalanceOrg"] + epsilon)
    )
    out["dest_balance_change"] = out["newbalanceDest"] - out["oldbalanceDest"]

    # --- transaction type encoding --------------------------------------
    # drop_first=True drops CASH_IN as the reference category. With dtype=int
    # the columns are guaranteed numeric for the scaler.
    out = pd.get_dummies(out, columns=["type"], drop_first=True, dtype=int)

    # --- drop raw columns now superseded by engineered features ---------
    raw_columns_to_drop = [
        "amount",
        "oldbalanceOrg", "newbalanceOrig",
        "oldbalanceDest", "newbalanceDest",
    ]
    out = out.drop(columns=raw_columns_to_drop)

    return out


In [5]:
X_train_fe = build_features(X_train)
X_test_fe = build_features(X_test)

print(f"Train: {X_train_fe.shape}")
print(f"Test:  {X_test_fe.shape}")
print(f"Feature columns: {list(X_train_fe.columns)}")


Train: (5090096, 11)
Test:  (1272524, 11)
Feature columns: ['step', 'log_amount', 'abs_balance_error', 'log_balance_error', 'log_amount_balance_error', 'log_amount_to_balance_ratio', 'dest_balance_change', 'type_CASH_OUT', 'type_DEBIT', 'type_PAYMENT', 'type_TRANSFER']


## 3. Sanity checks

A subtle point worth checking: if a transaction type appears in the training set but not the test set (or vice versa), one-hot encoding produces different column sets and the modelling step will fail at predict time. PaySim has only five types and all are heavily represented, so this won't happen here. The FastAPI service sees one transaction at a time and will need to handle missing dummies explicitly, we address that in `src/api/main.py`.

In [6]:
# Column sets must match exactly between train and test. If they don't, the
# scaler and the models will silently misalign features at scoring time.
assert list(X_train_fe.columns) == list(X_test_fe.columns), (
    "Train and test feature columns differ — one-hot encoding produced "
    "different sets. Inspect df['type'].value_counts() in each split."
)

# No NaN or infinite values should reach the modelling step. Both would
# happen silently if division-by-zero or log of negative leaked through.
assert X_train_fe.notna().all().all(), "NaN in training features"
assert X_test_fe.notna().all().all(), "NaN in test features"
assert np.isfinite(X_train_fe.values).all(), "Infinite values in training features"
assert np.isfinite(X_test_fe.values).all(), "Infinite values in test features"

print("Column alignment and finiteness checks passed.")

Column alignment and finiteness checks passed.


## 4. Scaling: StandardScaler vs RobustScaler

`StandardScaler` rescales each feature by `(x − mean) / std`. The mean and standard deviation are both highly sensitive to outliers, which makes `StandardScaler` a poor fit for heavy-tailed financial variables — a single transaction at 92 million dollars dominates both statistics and pulls the rest of the distribution toward zero.

`RobustScaler` uses `(x − median) / IQR` instead. The median and interquartile range are robust to outliers, so the resulting scale reflects the bulk of the distribution rather than its tail. This is the more defensible default for transaction amounts, balance changes, and any other heavy-tailed monetary variable.

We compare the two below and use `RobustScaler` for the production pipeline. Tree-based models are scale-invariant and would not care either way, but the autoencoder in notebook 04 and any distance-based detector are sensitive to feature scale, and the comparison is a methodological signal that the choice of scaler was deliberate rather than default.


In [7]:
skew_diagnostic = X_train_fe.skew().sort_values(ascending=False)
print("Feature skewness (sorted, descending):")
print(skew_diagnostic.round(2))

Feature skewness (sorted, descending):
dest_balance_change            32.53
abs_balance_error              29.60
type_DEBIT                     12.27
type_TRANSFER                   3.00
log_amount_to_balance_ratio     0.70
type_PAYMENT                    0.68
type_CASH_OUT                   0.62
step                            0.38
log_amount                     -0.56
log_balance_error              -1.06
log_amount_balance_error       -1.10
dtype: float64


The features with the heaviest tails are exactly the ones we expect `dest_balance_change` (raw monetary change at the receiver) and the residual-derived features that haven't been log-transformed in their absolute form. These are the features for which the choice between `StandardScaler` and `RobustScaler` matters most.


In [ ]:
# Fit on train, apply to both train and test. 
scaler = RobustScaler()

X_train_scaled = pd.DataFrame(
    scaler.fit_transform(X_train_fe),
    columns=X_train_fe.columns,
    index=X_train_fe.index,
)
X_test_scaled = pd.DataFrame(
    scaler.transform(X_test_fe),
    columns=X_test_fe.columns,
    index=X_test_fe.index,
)

# Sanity check: 
sanity = pd.DataFrame({
    "train_median": X_train_scaled.median(),
    "train_iqr":    X_train_scaled.quantile(0.75) - X_train_scaled.quantile(0.25),
    "test_median":  X_test_scaled.median(),
}).round(4)
sanity

,train_median,train_iqr,test_median
step,0.0,1.0,0.0000
log_amount,0.0,1.0,0.0008
abs_balance_error,0.0,1.0,0.0016
log_balance_error,-0.0,1.0,0.0013
log_amount_balance_error,0.0,1.0,0.0014
log_amount_to_balance_ratio,0.0,1.0,-0.0000
dest_balance_change,0.0,1.0,0.0000
type_CASH_OUT,0.0,1.0,0.0000
type_DEBIT,0.0,0.0,0.0000
type_PAYMENT,0.0,1.0,0.0000


Training-set medians are all 0.0 and IQRs are all 1.0, which is the expected result of RobustScaler by construction. Test medians are close but not exactly zero which is correct, since the test set was scaled using training statistics rather than its own. The two binary type columns (`type_DEBIT`, `type_TRANSFER`) show IQR = 0.0 because they are sparse indicators; this does not affect tree-based models, which are scale-invariant.

## 5. Persisting artefacts

Three files are written:

- `data/processed/train.csv` and `data/processed/test.csv` — scaled features plus the target, used by notebooks 03–05.
- `src/models/scaler.pkl` — the fitted scaler, loaded by the FastAPI service so a single incoming transaction goes through the same transformation as training did.
- `src/models/feature_columns.pkl`, the column order, used by the API to align a single-row dataframe to the model's expected input.

CSV is chosen over parquet for portability with the API prototype and ease of inspection. For a real deployment with billions of rows, parquet plus a feature store would replace this.

In [14]:
# Re-attach the target so each split is a single self-contained file.
train_out = X_train_scaled.assign(isFraud=y_train.values)
test_out = X_test_scaled.assign(isFraud=y_test.values)

train_out.to_csv(PROCESSED_DIR / "train.csv", index=False)
test_out.to_csv(PROCESSED_DIR / "test.csv", index=False)

joblib.dump(scaler, MODELS_DIR / "scaler.pkl")
joblib.dump(list(X_train_fe.columns), MODELS_DIR / "feature_columns.pkl")

print("Wrote:")
for p in [
    PROCESSED_DIR / "train.csv",
    PROCESSED_DIR / "test.csv",
    MODELS_DIR / "scaler.pkl",
    MODELS_DIR / "feature_columns.pkl",
]:
    print(f"  {p}")


Wrote:
  ..\data\processed\train.csv
  ..\data\processed\test.csv
  ..\src\models\scaler.pkl
  ..\src\models\feature_columns.pkl


## 6. Summary

The training set has 5.09M rows and the test set 1.27M, both stratified to preserve the 0.13% fraud rate. Each transaction is represented by 11 features: `log_amount`, `abs_balance_error`, `log_balance_error`, 
`log_amount_balance_error`, `log_amount_to_balance_ratio`, `dest_balance_change`, and four type indicators (`type_CASH_OUT`, `type_DEBIT`, `type_PAYMENT`, `type_TRANSFER`, with `CASH_IN` as the reference category).

Three methodological choices in this notebook go beyond the original feature list. The train/test split happens before any preprocessing to prevent leakage. RobustScaler replaces StandardScaler because skewness 
above 29 in two features makes mean and standard deviation unreliable statistics. 

The next notebook trains the supervised baselines on this feature set.